File: SRM 674/FindingKids.java
import java.io.*;
import java.util.*;
import java.math.*;

public class FindingKids {
    final int MOD = 1000000007;
	public long getSum(int n, int q, int A, int B, int C) {
        Set<Integer> locs = new HashSet<Integer>();
        long curr = A;
        State[] list = new State[n];
        for (int i = 0; i < n; ++i) {
            curr *= B;
            curr += C;
            curr %= MOD;
            int p = (int) curr;
            p %= (MOD - n + i + 1);
            if (!locs.add(p)) {
                p = MOD - n + i;
            }
            locs.add(p);
            list[i] = new State(i, p, p % 2 != 0);
        }
        int[] order = new int[n];
        Arrays.sort(list);
        ArrayList<Integer> inc = new ArrayList<Integer>();
        ArrayList<Integer> dec = new ArrayList<Integer>();
        for (int i = 0; i < order.length; ++i) {
            order[list[i].kidIndex] = i;
            if (list[i].l) {
                dec.add(list[i].x);
            } else {
                inc.add(list[i].x);
            }
        }
        long ret = 0;
        int[] kid = new int[q];
        int[] time = new int[q];
        for (int i = 0; i < q; ++i) {
            curr *= B;
            curr += C;
            curr %= MOD;
            kid[i] = (int) (curr%n);
            curr *= B;
            curr += C;
            curr %= MOD;
            time[i] = (int) curr;
        }
        for (int i = 0; i < q; ++i) {
            ret = Math.abs(solve(inc, dec, order[kid[i]] + 1, time[i]));

        }
        return ret;

	}

    public static long solve(ArrayList<Integer> inc, ArrayList<Integer> dec, int indexWant, int time) {
        int l = 0;
        int r = inc.size() - 1;
        while (l <= r) {
            int m = (l + r) / 2;
            long goal = inc.get(m) + time;
            int other1 = howManyLess(dec, - 1, goal, time);
            int other2 = howManyLessEq(dec, -1, goal, time);
            if (m + other1 <= indexWant - 1 && m + other2 >= indexWant - 1) {
                return goal;
            }
            else if (m + other1 >= indexWant) {
                r = m - 1;
            }
            else {
                l = m + 1;
            }

        }
        l = 0;
        r = dec.size() - 1;
        while (l <= r) {
            int m = (l + r) / 2;
            long goal = dec.get(m) - time;
            int other1 = howManyLess(inc, 1, goal, time);
            int other2 = howManyLessEq(inc, 1, goal, time);
            if (m + other1 <= indexWant - 1 && m + other2 >= indexWant - 1) {
                return goal;
            }
            else if (m + other1 >= indexWant) {
                return m - 1;
            }
            else {
                l = m + 1;
            }
        }
        throw new RuntimeException();
    }


    public static int howManyLess(ArrayList<Integer> list, int scale, long goal, int time) {
        if (list.isEmpty()) return 0;
        if (list.get(0) + scale * time >= goal) return 0;
        int l = 0;
        int r = list.size() - 1;
        while (l != r) {
            int m = (l + r + 1) / 2;
            if (list.get(m) + scale * time >= goal) {
                r = m - 1;
            } else {
                l = m;
            }
        }
        return l + 1;

    }

    public static int howManyLessEq(ArrayList<Integer> list, int scale, long goal, int time) {
        if (list.isEmpty()) return 0;
        if (list.get(0) + scale * time > goal) return 0;
        int l = 0;
        int r = list.size() - 1;
        while (l != r) {
            int m = (l + r + 1) / 2;
            if (list.get(m) + scale * time > goal) {
                r = m - 1;
            } else {
                l = m;
            }
        }
        return l + 1;
    }

    static class State implements Comparable<State> {
        public int kidIndex;
        public int x;
        public boolean l;
        public State(int kidIndex, int x, boolean l) {
            super();
            this.kidIndex = kidIndex;
            this.x = x;
            this.l = l;
        }
        public int compareTo(State s) {
            return x - s.x;
        }
    }

// CUT begin
	public static void main(String[] args){
		System.err.println("FindingKids (500 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int n = Integer.parseInt(Reader.nextLine());
            int q = Integer.parseInt(Reader.nextLine());
            int A = Integer.parseInt(Reader.nextLine());
            int B = Integer.parseInt(Reader.nextLine());
            int C = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            long __answer = Long.parseLong(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(n, q, A, B, C, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1483168588;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int n, int q, int A, int B, int C, long __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		FindingKids instance = new FindingKids();
		long __result = 0;
		try {
			__result = instance.getSum(n, q, A, B, C);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "FindingKids.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: SRM 674/FindingKids.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>
#include <map>
#include <algorithm>
#define mp make_pair
#define pb push_back      
#define F first
#define S second
#define SS stringstream
#define sqr(x) ((x)*(x))
#define m0(x) memset(x,0,sizeof(x))
#define m1(x) memset(x,63,sizeof(x))
#define CC(x) cout << (x) << endl
#define pw(x) (1ll<<(x))
#define buli(x) __builtin_popcountll(x)
#define M 1000000007
#define N 2000111
#define rep(i,n) for(int i=0;i<n;i++)

using namespace std;
 
int pos[N];
int dir[N];
int kid[N], ti[N];
 
long long ans[N];
int n, q;
 
void sol() {
    vector<int> L, R;
    rep(i,n) if(0 == dir[i]) L.pb(pos[i]); else R.pb(pos[i]);
    sort(L.begin(), L.end());
    sort(R.begin(), R.end());
    for (int i = 0; i < q; i++) {
        if (1 == dir[kid[i]]) {
            int x = lower_bound(R.begin(), R.end(), pos[kid[i]]) - R.begin();
            int y = lower_bound(L.begin(), L.end(), pos[kid[i]]) - L.begin();
            int l = 0;
            int r = n + 1;
            while (l < r) {
                int mid = (l + r + 1) / 2;
                int xx = x - mid / 2;
                int yy = y + (mid + 1) / 2- 1;
                if (xx < 0 || yy >= L.size() || abs(R[xx] - L[yy]) / 2. > ti[i]) r = mid - 1; else l = mid;
            }
            if (l % 2 == 0) {
                ans[i] = abs(R[x - l / 2] + ti[i]);
            } else {
                ans[i] = abs(L[y + (l + 1) / 2 - 1] - ti[i]);
            }
        }
    }
}

class FindingKids {
    public:
    long long getSum(int nn, int qq, int A, int B, int C) {
        n = nn;
        q = qq;
        set<int> was;
        rep(i, n) {
            A = (A * 1ll * B + C) % M;
            int p = A % (M - n + i + 1);
            if (was.find(p) != was.end()) {
                p = M - n + i;
            }
            pos[i] = p;
            was.insert(p);
            if (p % 2 == 0) dir[i] = 1; else dir[i] = 0;
        }
        rep(i, q) {
            A = (A * 1ll * B + C) % M;
            kid[i] = A % n;
            A = (A * 1ll * B + C) % M;
            ti[i] = A;
        }
        sol();
        for (int i = 0; i < n; i++) {
            pos[i] *= -1;
            dir[i] ^= 1;
        }
        sol();
        long long ret = 0;
        rep(i, q) ret += ans[i];
        return ret;
  }
};

// CUT begin
ifstream data("FindingKids.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(int n, int q, int A, int B, int C, long long __expected) {
    time_t startClock = clock();
    FindingKids *instance = new FindingKids();
    long long __result = instance->getSum(n, q, A, B, C);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        int n;
        from_stream(n);
        int q;
        from_stream(q);
        int A;
        from_stream(A);
        int B;
        from_stream(B);
        int C;
        from_stream(C);
        next_line();
        long long __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(n, q, A, B, C, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1483146196;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "FindingKids (500 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
